miio/gateway.py

```diff
@@ -536,38 +513,216 @@ class SubDevice:
+        """Returns firmware version"""
+        return self.get_subdevice_prop("fw_ver").pop()
+
+
+class AqaraHT(SubDevice):
+    _temperature = None
+    _humidity = None
+    _pressure = None
+
+    @property
+    def temperature(self):
+        """Return the temperature in degrees celsius"""
+        return self._temperature
+
+    @property
+    def humidity(self):
+        """Return the humidity in %"""
+        return self._humidity
+
+    @property
+    def pressure(self):
+        """Return the pressure in hPa"""
+        return self._pressure
+
+    @command()
+    def update(self):
+        """Update all device properties"""
+        values = self.get_subdevice_prop_exp(["temperature", "humidity", "pressure"])
+        if len(values) == 3:
+            self._temperature = values[0]/100
+            self._humidity = values[1]/100
+            self._pressure = values[2]/100
+        return
 
 
 class SensorHT(SubDevice):
     accessor = "get_prop_sensor_ht"
     properties = ["temperature", "humidity"]
+    _temperature = None
+    _humidity = None
 
+    @property
+    def temperature(self):
+        return self._temperature
+
+    @property
+    def humidity(self):
+        return self._humidity
+    
+    @command()
+    def update(self):
+        values = self.get_subdevice_prop_exp(properties)
+        if len(values) == 2:
+            self._temperature = values[0]
+            self._humidity = values[1]
+        return
 
 class Plug(SubDevice):
     accessor = "get_prop_plug"
     properties = ["power", "neutral_0"]
+    _power = None
+    _status = None
+
+    @property
+    def power(self):
+        return self._power
+
+    @property
+    def status(self):
+        return self._status
+    
+    @command()
+    def update(self):
+        values = self.get_subdevice_prop_exp(properties)
+        if len(values) == 2:
+            self._power = values[0]
+            self._status = values[1]
+        return
+
+class AqaraRelayTwoChannels(SubDevice):
+    _status_ch0 = None
+    _status_ch1 = None
+    _load_power = None
+
+    class AqaraRelayToggleValue(Enum):
+        toggle = "toggle"
+        on = "on"
+        off = "off"
+
+    class AqaraRelayChannel(Enum):
+        first = "channel_0"
+        second = "channel_1"
+
+    @property
+    def status_ch0(self):
+        """Return the state of channel 0"""
+        return self._status_ch0
+
+    @property
+    def status_ch1(self):
+        """Return the state of channel 1"""
+        return self._status_ch1
+
+    @property
+    def load_power(self):
+        """Return the load power"""
+        return self._load_power
+
+    @command()
+    def update(self):
+        """Update all device properties"""
+        values = self.get_subdevice_prop_exp(["load_power", "channel_0", "channel_1"])
+        if len(values) == 3:
+            self._load_power = values[0]
+            self._status_ch0 = values[1]
+            self._status_ch1 = values[2]
+        return
```

Comment:
See the above comment about error handling.

Reply:
done